### PR TITLE
[profiling] add ffi for crossbeam::ArrayQueue

### PR DESCRIPTION
**Pull request by taegyunkim**
# What does this PR do?

Adds ffi for `crossbeam::ArrayQueue<*mut c_void>`

# Motivation

Python profiler allocates a new `Sample` whenever a it collects an event (stack/memory/lock).   Then, the Sample is freed right after. Code for `new/delete Sample` is [here](https://github.com/DataDog/dd-trace-py/blob/d1db2005a8e301816f7618ee2b14ab9acfec3185/ddtrace/internal/datadog/profiling/dd_wrapper/src/sample_manager.cpp#L30-L40). This results in calling `new/delete` a lot and creates some memory pressure to the system and shows up on native profiler. Frequent calls to `new/delete` can be avoided if we use a pool of `Sample`s. And this `ArrayQueue` will be used to implement the pool. 

Another option we considered is using mutex guarded vector in C++ in dd-trace-py. However, it would be little tricky to handle `fork()` if we use a lock. So we decided to use lock-free list/queue/stack.

# Additional Notes

Anything else we should know when reviewing?

# How to test the change?

Describe here in detail how the change can be validated.

### CODE REVIEW

File: profiling-ffi/src/lib.rs
Comment (danielsn) [inline profiling-ffi/src/lib.rs:12]: Would this make sense in `ddcommon-ffi` since there is nothing particularly profiling about it?

File: ddcommon-ffi/src/array_queue.rs
Comment (danielsn) [inline ddcommon-ffi/src/array_queue.rs:40]: safety comment

File: ddcommon-ffi/src/array_queue.rs
Comment (danielsn) [inline ddcommon-ffi/src/array_queue.rs:81]: safety comment

File: ddcommon-ffi/src/array_queue.rs
Comment (ivoanjo) [inline ddcommon-ffi/src/array_queue.rs:71]: Hope you don't mind the tiny nitpick, but for consistency the naming we've adopted calls tries to match the rust struct name on the C side, e.g. 
`ArrayQueue.new` => `ddog_ArrayQueue_new` :)

(This applies to the other public functions as well)

File: ddcommon-ffi/src/array_queue.rs
Comment (ivoanjo) [inline ddcommon-ffi/src/array_queue.rs:160]: This comment seems to have been cut-off?

File: ddcommon-ffi/src/array_queue.rs
Comment (ivoanjo) [inline ddcommon-ffi/src/array_queue.rs:112]: Minor: It took me a while to understand why it was useful to return the value back, since for `push` it's somewhat redundant. Maybe include here a comment saying returning a value is mostly for `force_push`?

**Review by ivoanjo — APPROVED** [review]
This looks quite reasonable! I've been holding off on buffering things on the Ruby library (instead try to make everything immediate) in part because of the complexity of getting this kind of queue going in a C codebase, but I guess after this is in I should probably consider using it in a few places :)

My rust-foo isn't great so I recommend waiting for @danielsn's final pass/approval as well, but here sir, take my :+1:

File: ddcommon-ffi/src/array_queue.rs
Comment (paullegranddc) [inline ddcommon-ffi/src/array_queue.rs:84]: If this function is not exported (it cannot really since it returns non C types) it probably shouldn't follow the `ddog_array_queue` prefix, because it's not very idiomatic from the rust side.

Maybe make it an associated function on the type?
```rust
impl ArrayQueue {
  fn into_inner(queue_ptr: *mut ArrayQueue) -> anyhow::Result<&'a mut crossbeam_queue::ArrayQueue<*mut c_void>> {
}

// then call like
let queue = ArrayQueue::into_inner(queue_ptr);
```
}

**Review by paullegranddc — COMMENTED** [review]
As a general comment, all of the `extern "C"` functions in this PR take an `*ArrayQueue`  and cast it to an `&'a crossbeam::ArrayQueue`,  but the thing is, the lifetime is never bound, so you could very well get an `&'static crossbeam::ArrayQueue` and have it escape the fucntion.


I think it'd be better to take an `&'a ArrayQueue` in the ffi fucntions, which is translated to the same `ArrayQueue*` in the C header, which would allow to derive the `&'a crossbeam::ArrayQueue` lifetime safely and ensure it doesn't escape.

File: ddcommon-ffi/src/array_queue.rs
Comment (paullegranddc) [reply ddcommon-ffi/src/array_queue.rs:84]: Also, this function casts the ArrayQueue to an `&mut`. So, it's undefined behavior to call it concurrently. Most functions on the crossbeam::ArrayQueue only require &self anyway, so IMO you should probably have to cast functions, one creating a ref for push/pop and another one getting a ref_mut for the drop.

Comment (taegyunkim) [reply ddcommon-ffi/src/array_queue.rs:84]: Done. see `as_inner_ref()` and `take()`

File: ddcommon-ffi/src/array_queue.rs
Comment (taegyunkim) [inline ddcommon-ffi/src/array_queue.rs:33]: To reviewers, @ivoanjo @danielsn @paullegranddc 

Not sure whether it would be clearer to have `take()` or put this in `drop()`. Any suggestions?

File: examples/ffi/array_queue.cpp
Comment (taegyunkim) [inline examples/ffi/array_queue.cpp:1]: I've only tested running this with `valgrind`(`leaks` on Mac). Any other suggestions for testing?

File: ddcommon-ffi/src/array_queue.rs
Comment (danielsn) [inline ddcommon-ffi/src/array_queue.rs:16]: Can this be null?

File: ddcommon-ffi/src/array_queue.rs
Comment (danielsn) [inline ddcommon-ffi/src/array_queue.rs:15]: Can this be null?

File: ddcommon-ffi/src/array_queue.rs
Comment (danielsn) [inline ddcommon-ffi/src/array_queue.rs:23]: This should return `Result` and check if item_delete_fn is null

File: ddcommon-ffi/src/array_queue.rs
Comment (danielsn) [inline ddcommon-ffi/src/array_queue.rs:21]: 0 capacity is illegal

File: ddcommon-ffi/src/array_queue.rs
Comment (danielsn) [reply ddcommon-ffi/src/array_queue.rs:33]: I'd put it in `drop`.  Having `take` do anything other than just replacing the value with default violates principal of least suprise.

File: ddcommon-ffi/src/array_queue.rs
Comment (danielsn) [inline ddcommon-ffi/src/array_queue.rs:15]: I wonder if we can just make this a `Box<ArrayQueue>` https://github.com/rust-lang/unsafe-code-guidelines/issues/157

File: ddcommon-ffi/src/array_queue.rs
Comment (danielsn) [inline ddcommon-ffi/src/array_queue.rs:11]: Since you implement drop, you might need a phantom data

File: ddcommon-ffi/src/array_queue.rs
Comment (danielsn) [inline ddcommon-ffi/src/array_queue.rs:97]: do we want to alert on the case of double free?

File: ddcommon-ffi/src/array_queue.rs
Comment (danielsn) [inline ddcommon-ffi/src/array_queue.rs:125]: describe the return value

File: examples/ffi/array_queue.cpp
Comment (danielsn) [reply examples/ffi/array_queue.cpp:1]: fuzzing?

**Comment by gleocadie** [comment]
Can you add description and what motivates this PR please ? (as describe in the template)
It's difficult to understand why we are adding this and if it would benefit us (other libraries)
Thx

File: ddcommon-ffi/src/array_queue.rs
Comment (taegyunkim) [reply ddcommon-ffi/src/array_queue.rs:16]: Changed to use `NonNull`

File: ddcommon-ffi/src/array_queue.rs
Comment (taegyunkim) [reply ddcommon-ffi/src/array_queue.rs:15]: Changed to use `NonNull` with comments

Comment (taegyunkim) [reply ddcommon-ffi/src/array_queue.rs]: Changed to use `NonNull<ArrayQueue>`, thanks for pointing this out. And added a comment why `NonNull` instead of `Box`

File: ddcommon-ffi/src/array_queue.rs
Comment (taegyunkim) [reply ddcommon-ffi/src/array_queue.rs:12]: Would miri complain about this?

File: ddcommon-ffi/src/array_queue.rs
Comment (taegyunkim) [reply ddcommon-ffi/src/array_queue.rs]: Added comment

File: ddcommon-ffi/src/array_queue.rs
Comment (taegyunkim) [reply ddcommon-ffi/src/array_queue.rs:12]: Here the ArrayQueue only performs push and pop, don't have peek() which could have required me to have a phantom data.

File: ddcommon-ffi/src/array_queue.rs
Comment (taegyunkim) [reply ddcommon-ffi/src/array_queue.rs]: Also, changed the functions to take `&ArrayQueue` instead of `&mut`

File: examples/ffi/array_queue.cpp
Comment (taegyunkim) [reply examples/ffi/array_queue.cpp:1]: Added fuzz testing on rust side.

File: ddcommon-ffi/src/array_queue.rs
Comment (danielsn) [inline ddcommon-ffi/src/array_queue.rs:41]: This happens so rarely, it would be cleaner to just 
```suggestion
        let inner = NonNull::new(Box::into_raw(Box::new(internal_queue)) as *mut ArrayQueue)?;
```

File: ddcommon-ffi/src/array_queue.rs
Comment (danielsn) [inline ddcommon-ffi/src/array_queue.rs:272]: Use the same case as the function

File: ddcommon-ffi/src/array_queue.rs
Comment (danielsn) [inline ddcommon-ffi/src/array_queue.rs:434]: 💯

**Review by danielsn — APPROVED** [review]
Modulo a few  ⛏️  LGTM

File: ddcommon-ffi/src/array_queue.rs
Comment (danielsn) [inline ddcommon-ffi/src/array_queue.rs:36]:
```suggestion
```

There is no unsafe block anymore

**Pull request merged**
